proxy/coco/src/announce.rs

```diff
@@ -0,0 +1,303 @@
+//! Compute, track and announce noteworthy changes to the network.
+
+use std::collections::HashSet;
+use std::ops::Deref as _;
+
+use kv::Codec as _;
+
+use librad::keys;
+use librad::net;
+use librad::uri;
+
+use crate::error::Error;
+use crate::oid::Oid;
+use crate::peer;
+
+/// Name for the bucket used in [`kv::Store`].
+const BUCKET_NAME: &str = "announcements";
+/// Key for the single value used as cache.
+const KEY_NAME: &str = "latest";
+
+/// An update and all the required information that can be announced on the network.
+pub type Announcement = (uri::RadUrn, String, Oid);
+
+/// Announces the list of given `updates` with the [`librad::net::protocol`].
+///
+/// # Errors
+///
+/// * if the announcemnet of one of the project heads failed
+pub async fn announce(
+    protocol: net::protocol::Protocol<net::peer::PeerStorage<keys::SecretKey>, net::peer::Gossip>,
+    updates: Vec<Announcement>,
+) -> Result<(), Error> {
+    for (project_urn, head, hash) in &updates {
+        let urn = uri::RadUrn::new(
+            project_urn.id.clone(),
+            uri::Protocol::Git,
+            uri::Path::parse(head)?,
+        );
+        let have = net::peer::Gossip {
+            urn,
+            rev: Some(net::peer::Rev::Git((*hash).into())),
+            origin: Some(protocol.peer_id().clone()),
+        };
+
+        protocol.announce(have).await;
```

Comment:
Could be slow if lots of projects, perhaps something to run concurrently in the future..

Reply:
If for whatever reason announcing takes more than the tick rate, then you're gonna start to build up a bunch of latency

Reply:
It would only announce if there's a diff though right?

Reply:
Yeah true, and since you only have a diff on your own projects, it should be small enough.